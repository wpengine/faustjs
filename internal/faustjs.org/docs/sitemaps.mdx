---
slug: sitemaps
title: Sitemaps
description: Generate sitemaps for your Faust.js site
---

### This document is in progress. Check back soon for content!