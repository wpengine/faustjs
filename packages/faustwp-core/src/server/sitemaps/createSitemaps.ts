import type { IncomingMessage } from 'http';
import { X2jOptions, XMLParser } from 'fast-xml-parser';
import isArray from 'lodash/isArray.js';
import isUndefined from 'lodash/isUndefined.js';
import trim from 'lodash/trim.js';
import type { NextRequest } from 'next/server.js';
import { getWpUrl } from '../../lib/getWpUrl.js';
import {
  FAUST_PAGES_PATHNAME,
  GetSitemapPropsConfig,
  SITEMAP_INDEX_PATH,
} from './getSitemapProps.js';
import {
  createSitemap,
  createSitemapIndex,
  SitemapSchemaSitemapElement,
  SitemapSchemaUrlElement,
} from './sitemapUtils.js';

/**
 * TypeScript representation of a parsed XML sitemap
 */
export interface ParsedSitemap {
  urlset: {
    url: SitemapSchemaUrlElement[];
  };
}

/**
 * TypeScript representation of a parsed XML sitemap index
 */
export interface ParsedSitemapIndex {
  sitemapindex: {
    sitemap: SitemapSchemaSitemapElement[];
  };
}

const parserConfig: Partial<X2jOptions> = {
  ignoreAttributes: false,
  preserveOrder: false,
  unpairedTags: ['xml', 'xml-stylesheet'],
  processEntities: true,
  htmlEntities: true,
};

/**
 * Creates the root XML sitemap index (e.g. /sitemap.xml) that lists all the
 * sitemaps provided as the sitemapPaths property in the config, in addition to
 * a sitemap for the Next.js pages provided as the pages property in the config.
 *
 * @param {NextRequest} req The Next.js request object
 * @param {NormalizedConfig} normalizedConfig A normalized config object
 * @returns {Response|undefined}
 */
export async function createRootSitemapIndex(
  req: NextRequest | IncomingMessage,
  config: GetSitemapPropsConfig,
): Promise<Response | undefined> {
  const { pages, sitemapPathsToIgnore, frontendUrl, sitemapIndexPath } = config;

  if (!req.url) {
    throw new Error('Request object must have URL');
  }

  // get sitemapIndexPath config param
  // fetch sitemap from WP
  const trimmedWpUrl = trim(getWpUrl(), '/');
  const trimmedFrontendUrl = trim(frontendUrl, '/');
  const trimmedSitemapIndexPath = trim(
    sitemapIndexPath || SITEMAP_INDEX_PATH,
    '/',
  );
  const wpSitemapUrl = `${trimmedWpUrl}/${trimmedSitemapIndexPath}`;

  let sitemaps: SitemapSchemaSitemapElement[] = [];

  if (!isUndefined(pages) && isArray(pages) && pages.length) {
    const trimmedFaustPagesPart = `${trim(
      SITEMAP_INDEX_PATH,
      '/',
    )}?sitemap=${trim(FAUST_PAGES_PATHNAME, '/')}`;
    const sitemapFaustPagesUrl = `${trimmedFrontendUrl}/${trimmedFaustPagesPart}`;

    sitemaps = [
      ...sitemaps,
      {
        loc: encodeURI(sitemapFaustPagesUrl),
      },
    ];
  }

  const res = await fetch(wpSitemapUrl);

  // Don't proxy the sitemap index if the response was not ok.
  if (!res.ok) {
    console.error(
      `Failed to fetch sitemap index from WordPress at ${wpSitemapUrl}`,
    );
    console.error('Possible solutions:');
    console.error(
      '- Check that sitemapIndexPath is correct in the options passed to getSitemapProps. (some WordPress plugins change the default sitemap path)',
    );
    console.error(`- Consider flushing permalinks in WordPress.`);

    return undefined;
  }

  const xmlRes = await res.text();

  /**
   * Create a parser to convert our XML data into a JS object
   *
   * @link https://github.com/NaturalIntelligence/fast-xml-parser/blob/HEAD/docs/v4/6.HTMLParsing.md
   */
  const parser = new XMLParser({
    ...parserConfig,
    /**
     * FXP can not determine if a single tag should be parsed as an array or
     * an object, so we need to specify we always want "sitemap" tags to be an
     * array.
     *
     * @see https://github.com/NaturalIntelligence/fast-xml-parser/blob/master/docs/v4/2.XMLparseOptions.md#isarray
     */
    isArray: (tagName) => {
      return tagName === 'sitemap';
    },
  });

  // JS object representation of the XML sitemap index
  const parsedSitemapIndex: ParsedSitemapIndex = parser.parse(xmlRes);
  let wpSitemaps = parsedSitemapIndex?.sitemapindex?.sitemap;

  // The XML we parsed was not a proper sitemap
  if (isUndefined(wpSitemaps)) {
    return undefined;
  }

  /**
   * Ignore paths with exact matches to the sitemapPathsToIgnore property
   */
  wpSitemaps = wpSitemaps.filter((sitemap) => {
    const { pathname: sitemapPathname } = new URL(sitemap.loc);

    return !sitemapPathsToIgnore?.includes(sitemapPathname);
  });

  /**
   * Ignore paths from sitemapPathsToIgnore property that end in a wildcard
   */
  const wildcardPathsToIgnore = sitemapPathsToIgnore?.filter((path) =>
    path.endsWith('*'),
  );
  wpSitemaps = wpSitemaps.filter((sitemap) => {
    const { pathname: sitemapPathname } = new URL(sitemap.loc);

    let hasWildcard = false;

    wildcardPathsToIgnore?.forEach((path) => {
      const pathLessWildcard = path.slice(0, -1);
      if (sitemapPathname.startsWith(pathLessWildcard)) {
        hasWildcard = true;
      }
    });

    return !hasWildcard;
  });

  /**
   * Replace the existing WordPress URL in each "loc" with the frontend URL
   *
   * @example
   * Replaces http://headless.local/wp-sitemap-posts-page-1.xml with
   * http://localhost:3000/wp-sitemap-posts-page-1.xml
   */
  wpSitemaps.forEach((sitemap) => {
    const url = new URL(sitemap.loc);
    const sitemapUrl = `${trim(frontendUrl, '/')}/sitemap.xml?sitemap=${trim(
      url.pathname,
      '/',
    )}`;

    sitemaps = [
      ...sitemaps,
      {
        ...sitemap,
        loc: sitemapUrl,
      },
    ];
  });

  return createSitemapIndex(sitemaps);
}

/**
 * Creates a sitemap for the Next.js pages specified in the "pages" config option
 *
 * @param req The Next.js middleware request object
 * @param config The config object
 * @returns {Response|undefined}
 */
export function createPagesSitemap(
  req: NextRequest | IncomingMessage,
  config: GetSitemapPropsConfig,
): Response | undefined {
  if (!req.url) {
    throw new Error('Request object must have URL');
  }

  const { frontendUrl, pages } = config;

  if (isUndefined(pages) || !isArray(pages) || !pages.length) {
    return undefined;
  }

  let urls: SitemapSchemaUrlElement[] = [];

  pages.forEach((page) => {
    urls = [
      ...urls,
      {
        loc: `${trim(frontendUrl, '/')}/${trim(page.path, '/')}`,
        lastmod: page?.lastmod,
        changefreq: page?.changefreq,
        priority: page?.priority,
      },
    ];
  });

  return createSitemap(urls);
}

/**
 * Handles a request to a sitemap path listed in the sitemapPaths config option
 *
 * @param req The Next.js middleware request object
 * @param config The config object
 * @returns {Promise<Response|Undefined>}
 */
export async function handleSitemapPath(
  req: NextRequest | IncomingMessage,
  config: GetSitemapPropsConfig,
): Promise<Response | undefined> {
  if (!req.url) {
    throw new Error('Request object must have URL');
  }

  const { frontendUrl } = config;
  const paramsIndex = req.url.indexOf('?');
  const searchParamString = req.url.substr(paramsIndex);
  const urlParams = new URLSearchParams(searchParamString);

  const sitemapPath = urlParams.get('sitemap') as string;

  const wpSitemapUrl = `${trim(getWpUrl(), '/')}/${trim(sitemapPath, '/')}`;

  const res = await fetch(wpSitemapUrl);

  // Don't proxy the sitemap if the response was not ok.
  if (!res.ok) {
    return undefined;
  }

  const xmlRes = await res.text();

  /**
   * Create a parser to convert our XML data into a JS object
   *
   * @link https://github.com/NaturalIntelligence/fast-xml-parser/blob/HEAD/docs/v4/6.HTMLParsing.md
   */
  const parser = new XMLParser({
    ...parserConfig,
    /**
     * FXP can not determine if a single tag should be parsed as an array or
     * an object, so we need to specify we always want "url" tags to be an
     * array.
     *
     * @see https://github.com/NaturalIntelligence/fast-xml-parser/blob/master/docs/v4/2.XMLparseOptions.md#isarray
     */
    isArray: (tagName) => {
      return tagName === 'url';
    },
  });

  // JS object representation of the XML sitemap
  const parsedSitemap: ParsedSitemap = parser.parse(xmlRes);
  const wpSitemapUrls = parsedSitemap?.urlset?.url;

  // The XML we parsed was not a proper sitemap
  if (isUndefined(wpSitemapUrls)) {
    return undefined;
  }

  let urls: SitemapSchemaUrlElement[] = [];

  /**
   * Replace the existing WordPress URL in each "loc" with the headless URL
   *
   * @example
   * Replaces http://headless.local/wp-sitemap-posts-page-1.xml with
   * http://localhost:3000/wp-sitemap-posts-page-1.xml
   */
  wpSitemapUrls.forEach((url) => {
    urls = [
      ...urls,
      {
        ...url,
        loc: url.loc.replace(trim(getWpUrl(), '/'), trim(frontendUrl, '/')),
      },
    ];
  });

  return createSitemap(urls);
}
