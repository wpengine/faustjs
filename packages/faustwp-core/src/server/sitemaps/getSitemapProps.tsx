import { GetServerSidePropsContext } from 'next';
import isUndefined from 'lodash/isUndefined.js';
import isString from 'lodash/isString.js';
import isArray from 'lodash/isArray.js';
import isObject from 'lodash/isObject.js';
import { getQueryParam } from '../../utils/convert.js';
import {
  createPagesSitemap,
  createRootSitemapIndex,
  handleSitemapPath,
} from './createSitemaps.js';
import { SitemapSchemaUrlElement } from './sitemapUtils.js';

/**
 * The pathname to the root sitemap index.
 */
export const SITEMAP_INDEX_PATH = '/sitemap.xml';

/**
 * The pathname to the Next.js pages sitemap file. We may want to make this
 * configurable in the future.
 */
export const FAUST_PAGES_PATHNAME = '/sitemap-faust-pages.xml';

/**
 * the "pages" config type
 */
export interface NextJSPage extends Omit<SitemapSchemaUrlElement, 'loc'> {
  /**
   * The relative URL of the Next.js page.
   *
   * @example /about
   */
  path: string;
}

export type GetSitemapPropsConfig = {
  /**
   * The URL to your Faust site
   */
  frontendUrl: string;
  /**
   * A list of pathnames to ignore when proxying sitemaps.
   */
  sitemapPathsToIgnore?: string[];
  /**
   * The path to the sitemap index file in WordPress.
   */
  sitemapIndexPath?: string;
  /**
   * Next.js pages you want included in you sitemap. When provided, an index
   * will be created specifically for these pages.
   */
  pages?: NextJSPage[];
};

/**
 * Validates the structure of the user defined config.
 *
 * @param {Partial<HandleSitemapRequestsConfig>} config The user provided config
 */
export function validateConfig(config: Partial<GetSitemapPropsConfig>): void {
  if (!isString(config?.frontendUrl)) {
    throw new Error('frontendUrl must be a string');
  }

  try {
    // eslint-disable-next-line no-new
    const url = new URL(config.frontendUrl);

    if (url.protocol !== 'http:' && url.protocol !== 'https:') {
      throw new Error('URL must have protocol');
    }
  } catch (e) {
    throw new Error('frontendUrl must be a valid URL.');
  }

  if (!isUndefined(config?.sitemapPathsToIgnore)) {
    if (!isArray(config.sitemapPathsToIgnore)) {
      throw new Error('sitemapPathsToIgnore must be an array');
    }

    config?.sitemapPathsToIgnore?.forEach((path) => {
      if (!isString(path)) {
        throw new Error('sitemapPathsToIgnore must be an array of strings');
      }

      if (!path.startsWith('/')) {
        throw new Error(
          'Each sitemapPathsToIgnore must start with a forward slash',
        );
      }

      if (path.includes('*') && !path.endsWith('*')) {
        throw new Error(
          'sitemapPathsToIgnore with a wildcard must end with a wildcard',
        );
      }
    });
  }

  // Validate pages structure and required values
  if (!isUndefined(config?.pages)) {
    if (!isArray(config.pages)) {
      throw new Error('pages must be an array');
    }

    config?.pages?.forEach((page) => {
      if (!isObject(page)) {
        throw new Error('pages must be an array of objects');
      }

      if (isUndefined(page.path)) {
        throw new Error('pages must have a path property');
      }

      if (!isString(page.path)) {
        throw new Error('The pages path property must be a string');
      }
    });
  }
}

export async function getSitemapProps(
  ctx: GetServerSidePropsContext,
  config: GetSitemapPropsConfig,
) {
  // config validation with middleware flag
  validateConfig(config);

  if (!ctx.req.url) {
    throw new Error('A context url is required.');
  }

  const queryParam = getQueryParam(ctx.req.url, 'sitemap');

  if (queryParam === '') {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
    const response = await createRootSitemapIndex(ctx.req, config);
    if (!response || response?.status === 404) {
      return {
        notFound: true,
      };
    }
    ctx.res.setHeader('Content-Type', 'application/xml');

    ctx.res.write(await response?.text());

    ctx.res.end();
  }

  if (queryParam !== '' && queryParam !== 'sitemap-faust-pages.xml') {
    const response = await handleSitemapPath(ctx.req, config);

    if (!response || response?.status === 404) {
      return {
        notFound: true,
      };
    }
    ctx.res.setHeader('Content-Type', 'application/xml');

    ctx.res.write(await response?.text());
    ctx.res.end();
  }

  if (queryParam !== '' && queryParam === 'sitemap-faust-pages.xml') {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
    const response = createPagesSitemap(ctx.req, config);

    ctx.res.setHeader('Content-Type', 'application/xml');

    ctx.res.write(await response?.text());

    ctx.res.end();
  }

  return {
    props: {},
  };
}
