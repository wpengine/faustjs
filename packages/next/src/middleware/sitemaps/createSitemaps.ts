import { X2jOptions, XMLParser } from 'fast-xml-parser';
import { NextRequest } from 'next/server.js';
import {
  NormalizedConfig,
  FAUST_PAGES_PATHNAME,
} from './handleSitemapRequests.js';
import {
  createSitemap,
  createSitemapIndex,
  SitemapSchemaSitemapElement,
  SitemapSchemaUrlElement,
  isArray,
  trimSlashes,
  isUndefined,
} from './sitemapUtils.js';

/**
 * TypeScript representation of a parsed XML sitemap
 */
export interface ParsedSitemap {
  urlset: {
    url: SitemapSchemaUrlElement[];
  };
}

/**
 * TypeScript representation of a parsed XML sitemap index
 */
export interface ParsedSitemapIndex {
  sitemapindex: {
    sitemap: SitemapSchemaSitemapElement[];
  };
}

const parserConfig: Partial<X2jOptions> = {
  ignoreAttributes: false,
  preserveOrder: false,
  unpairedTags: ['xml', 'xml-stylesheet'],
  processEntities: true,
  htmlEntities: true,
};

/**
 * Creates the root XML sitemap index (e.g. /sitemap.xml) that lists all the
 * sitemaps provided as the sitemapPaths property in the config, in addition to
 * a sitemap for the Next.js pages provided as the pages property in the config.
 *
 * @param {NextRequest} req The Next.js middleware request object
 * @param {NormalizedConfig} normalizedConfig A normalized config object
 * @returns {Response|undefined}
 */
export async function createRootSitemapIndex(
  req: NextRequest,
  normalizedConfig: NormalizedConfig,
): Promise<Response | undefined> {
  const { pages, sitemapPathsToIgnore, replaceUrls, wpUrl } = normalizedConfig;
  const { pathname, origin } = new URL(req.url);
  let sitemaps: SitemapSchemaSitemapElement[] = [];

  if (!isUndefined(pages) && isArray(pages) && pages.length) {
    sitemaps = [
      ...sitemaps,
      { loc: `${trimSlashes(origin)}/${trimSlashes(FAUST_PAGES_PATHNAME)}` },
    ];
  }

  const wpSitemapUrl = `${trimSlashes(wpUrl)}/${trimSlashes(pathname)}`;
  const res = await fetch(wpSitemapUrl);

  // Don't proxy the sitemap index if the response was not ok.
  if (!res.ok) {
    return undefined;
  }

  const xmlRes = await res.text();

  /**
   * Create a parser to convert our XML data into a JS object
   *
   * @link https://github.com/NaturalIntelligence/fast-xml-parser/blob/HEAD/docs/v4/6.HTMLParsing.md
   */
  const parser = new XMLParser({
    ...parserConfig,
    /**
     * FXP can not determine if a single tag should be parsed as an array or
     * an object, so we need to specify we always want "sitemap" tags to be an
     * array.
     *
     * @see https://github.com/NaturalIntelligence/fast-xml-parser/blob/master/docs/v4/2.XMLparseOptions.md#isarray
     */
    isArray: (tagName) => {
      return tagName === 'sitemap';
    },
  });

  // JS object representation of the XML sitemap index
  const parsedSitemapIndex: ParsedSitemapIndex = parser.parse(xmlRes);
  let wpSitemaps = parsedSitemapIndex?.sitemapindex?.sitemap;

  // The XML we parsed was not a proper sitemap
  if (isUndefined(wpSitemaps)) {
    return undefined;
  }

  /**
   * Ignore paths with exact matches to the sitemapPathsToIgnore property
   */
  wpSitemaps = wpSitemaps.filter((sitemap) => {
    const { pathname: sitemapPathname } = new URL(sitemap.loc);

    return !sitemapPathsToIgnore?.includes(sitemapPathname);
  });

  /**
   * Ignore paths from sitemapPathsToIgnore property that end in a wildcard
   */
  const wildcardPathsToIgnore = sitemapPathsToIgnore?.filter((path) =>
    path.endsWith('*'),
  );
  wpSitemaps = wpSitemaps.filter((sitemap) => {
    const { pathname: sitemapPathname } = new URL(sitemap.loc);

    let hasWildcard = false;

    wildcardPathsToIgnore?.forEach((path) => {
      const pathLessWildcard = path.slice(0, -1);
      if (sitemapPathname.startsWith(pathLessWildcard)) {
        hasWildcard = true;
      }
    });

    return !hasWildcard;
  });

  /**
   * Replace the existing WordPress URL in each "loc" with the headless URL
   * if necessary
   *
   * @example
   * Replaces http://headless.local/wp-sitemap-posts-page-1.xml with
   * http://localhost:3000/wp-sitemap-posts-page-1.xml
   */
  if (replaceUrls) {
    wpSitemaps.forEach((sitemap) => {
      sitemaps = [
        ...sitemaps,
        {
          ...sitemap,
          loc: sitemap.loc.replace(trimSlashes(wpUrl), trimSlashes(origin)),
        },
      ];
    });
  } else {
    sitemaps = [...sitemaps, ...wpSitemaps];
  }

  return createSitemapIndex(sitemaps);
}

/**
 * Creates a sitemap for the Next.js pages specified in the "pages" config option
 *
 * @param req The Next.js middleware request object
 * @param normalizedConfig A normalized config object
 * @returns {Response|undefined}
 */
export function createPagesSitemap(
  req: NextRequest,
  normalizedConfig: NormalizedConfig,
): Response | undefined {
  const { origin } = new URL(req.url);
  const { pages } = normalizedConfig;

  if (isUndefined(pages) || !isArray(pages) || !pages.length) {
    return undefined;
  }

  let urls: SitemapSchemaUrlElement[] = [];

  pages.forEach((page) => {
    urls = [
      ...urls,
      {
        loc: `${trimSlashes(origin)}/${trimSlashes(page.path)}`,
        lastmod: page?.lastmod,
        changefreq: page?.changefreq,
        priority: page?.priority,
      },
    ];
  });

  return createSitemap(urls);
}

/**
 * Handles a request to a sitemap path listed in the sitemapPaths config option
 *
 * @param req The Next.js middleware request object
 * @param normalizedConfig A normalized config object
 * @returns {Promise<Response|Undefined>}
 */
export async function handleSitemapPath(
  req: NextRequest,
  normalizedConfig: NormalizedConfig,
): Promise<Response | undefined> {
  const { wpUrl, replaceUrls } = normalizedConfig;
  const { pathname, origin } = new URL(req.url);

  const wpSitemapUrl = `${trimSlashes(wpUrl)}/${trimSlashes(pathname)}`;
  const res = await fetch(wpSitemapUrl);

  // Don't proxy the sitemap if the response was not ok.
  if (!res.ok) {
    return undefined;
  }

  const xmlRes = await res.text();

  /**
   * Create a parser to convert our XML data into a JS object
   *
   * @link https://github.com/NaturalIntelligence/fast-xml-parser/blob/HEAD/docs/v4/6.HTMLParsing.md
   */
  const parser = new XMLParser({
    ...parserConfig,
    /**
     * FXP can not determine if a single tag should be parsed as an array or
     * an object, so we need to specify we always want "url" tags to be an
     * array.
     *
     * @see https://github.com/NaturalIntelligence/fast-xml-parser/blob/master/docs/v4/2.XMLparseOptions.md#isarray
     */
    isArray: (tagName) => {
      return tagName === 'url';
    },
  });

  // JS object representation of the XML sitemap
  const parsedSitemap: ParsedSitemap = parser.parse(xmlRes);
  const wpSitemapUrls = parsedSitemap?.urlset?.url;

  // The XML we parsed was not a proper sitemap
  if (isUndefined(wpSitemapUrls)) {
    return undefined;
  }

  let urls: SitemapSchemaUrlElement[] = [];

  if (replaceUrls) {
    /**
     * Replace the existing WordPress URL in each "loc" with the headless URL
     *
     * @example
     * Replaces http://headless.local/wp-sitemap-posts-page-1.xml with
     * http://localhost:3000/wp-sitemap-posts-page-1.xml
     */
    wpSitemapUrls.forEach((url) => {
      urls = [
        ...urls,
        {
          ...url,
          loc: url.loc.replace(trimSlashes(wpUrl), trimSlashes(origin)),
        },
      ];
    });
  } else {
    urls = wpSitemapUrls;
  }

  return createSitemap(urls);
}
