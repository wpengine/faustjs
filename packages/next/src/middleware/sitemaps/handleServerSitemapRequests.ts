import { NextRequest } from 'next/server.js';
import {
  createPagesSitemap,
  createRootSitemapIndex,
  handleSitemapPath,
  handleRobotsTxt,
} from './createSitemaps.js';
import {
  SitemapSchemaUrlElement,
  isArray,
  isBoolean,
  isString,
  isUndefined,
  isObject,
  isFunction,
} from './sitemapUtils.js';

/**
 * the "pages" config type
 */
export interface NextJSPage extends Omit<SitemapSchemaUrlElement, 'loc'> {
  /**
   * The relative URL of the Next.js page.
   *
   * @example /about
   */
  path: string;
}

export interface HandleSitemapRequestsConfig {
  /**
   * The URL of your WordPress site.
   *
   * @example https://my-wp-site.com
   */
  wpUrl: string;

  /**
   * The pathname to the sitemap index file.
   *
   * @example /sitemap.xml, /wp-sitemap.xml, /sitemap_index.xml
   */
  sitemapIndexPath: string;
  /**
   * A list of pathnames to ignore when proxying sitemaps.
   */
  sitemapPathsToIgnore?: string[];
  /**
   * Next.js pages you want included in you sitemap. When provided, an index
   * will be created specifically for these pages.
   */
  pages?: NextJSPage[];
  /**
   * Replace the WordPress site URL for your headless frontend site url in the
   * sitemap url entries.
   */
  replaceUrls?: boolean;
  /**
   * If defined, the text returned from this function will be used for the
   * `/robots.txt` route. The `/robots.txt` route will not be created if this
   * function is not defined.
   */
  robotsTxt?: (sitemapUrl: string) => Promise<string>;
}

export interface NormalizedConfig
  extends Omit<HandleSitemapRequestsConfig, 'replaceUrls'> {
  replaceUrls: boolean;
}

/**
 * The pathname to the Next.js pages sitemap file. We may want to make this
 * configurable in the future.
 */
export const FAUST_PAGES_PATHNAME = '/sitemap-faust-pages.xml';

/**
 * The pathname to the robots.txt file.
 */
export const FAUST_ROBOTS_PATHNAME = '/robots.txt';

/**
 * Validates the structure of the user defined config.
 *
 * @param {Partial<HandleSitemapRequestsConfig>} config The user provided config
 */
export function validateConfig(
  config: Partial<HandleSitemapRequestsConfig>,
): void {
  if (isUndefined(config?.wpUrl)) {
    throw new Error('wpUrl is required.');
  }

  if (!isString(config?.wpUrl)) {
    throw new Error('wpUrl must be a string.');
  }

  try {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const url = new URL(config.wpUrl);
  } catch (e) {
    throw new Error('wpUrl must be a valid URL.');
  }

  if (isUndefined(config?.sitemapIndexPath)) {
    throw new Error('sitemapIndexPath is required');
  }

  if (!isString(config?.sitemapIndexPath)) {
    throw new Error('sitemapIndexPath must be a string');
  }

  if (!config?.sitemapIndexPath.startsWith('/')) {
    throw new Error('sitemapIndexPath must start with a forward slash');
  }

  if (!isUndefined(config?.sitemapPathsToIgnore)) {
    if (!isArray(config.sitemapPathsToIgnore)) {
      throw new Error('sitemapPathsToIgnore must be an array');
    }

    config?.sitemapPathsToIgnore?.forEach((path) => {
      if (!isString(path)) {
        throw new Error('sitemapPathsToIgnore must be an array of strings');
      }

      if (!path.startsWith('/')) {
        throw new Error(
          'Each sitemapPathsToIgnore must start with a forward slash',
        );
      }

      if (path.includes('*') && !path.endsWith('*')) {
        throw new Error(
          'sitemapPathsToIgnore with a wildcard must end with a wildcard',
        );
      }
    });
  }

  // Validate pages structure and required values
  if (!isUndefined(config?.pages)) {
    if (!isArray(config.pages)) {
      throw new Error('pages must be an array');
    }

    config?.pages?.forEach((page) => {
      if (!isObject(page)) {
        throw new Error('pages must be an array of objects');
      }

      if (isUndefined(page.path)) {
        throw new Error('pages must have a path property');
      }

      if (!isString(page.path)) {
        throw new Error('The pages path property must be a string');
      }
    });
  }

  // Validate replaceUrls is a boolean
  if (!isUndefined(config?.replaceUrls) && !isBoolean(config?.replaceUrls)) {
    throw new Error('replaceUrls must be a boolean');
  }

  // Validate robotsTxt is a function
  if (!isUndefined(config?.robotsTxt) && !isFunction(config?.robotsTxt)) {
    throw new Error('robotsTxt must be a function');
  }
}

/**
 * Next.js middleware to proxy sitemap requests from the WordPress site.
 *
 * @param req The Next.js middleware request object
 * @param config The user specified config object
 * @returns {Response|undefined} A response object if the current request
 * is for a sitemap that needs to be handled, undefined otherwise
 */
export async function handleSitemapServerRequests(
  req: NextRequest,
  config: HandleSitemapRequestsConfig,
): Promise<Response | undefined> {
  // Validate config structure
  validateConfig(config);

  // Normalize config if some optional values are missing
  // eslint-disable-next-line prefer-object-spread
  const normalizedConfig: NormalizedConfig = Object.assign(
    {},
    { replaceUrls: true },
    config,
  ) as NormalizedConfig;

  const { pathname } = new URL(req.url);
  const { sitemapIndexPath, pages, robotsTxt } = normalizedConfig;

  // Handle the root XML sitemap if specified in the config
  if (pathname === sitemapIndexPath) {
    return createRootSitemapIndex(req, normalizedConfig);
  }

  // Handle the sitemap for the specified Next.js pages if specified in the config
  if (pathname === FAUST_PAGES_PATHNAME && pages?.length) {
    return createPagesSitemap(req, normalizedConfig);
  }

  // Handle the robots.txt file if specified in the config
  if (pathname === FAUST_ROBOTS_PATHNAME && robotsTxt) {
    return handleRobotsTxt(req, normalizedConfig);
  }

  // Handle the sitemap index paths specified in the config
  if (
    pathname.includes('sitemap') &&
    pathname.endsWith('.xml') &&
    !normalizedConfig?.sitemapPathsToIgnore?.includes(pathname)
  ) {
    return handleSitemapPath(req, normalizedConfig);
  }

  /**
   * If the above conditions are not met, return undefined and go on to the
   * next middleware
   */
  return undefined;
}
