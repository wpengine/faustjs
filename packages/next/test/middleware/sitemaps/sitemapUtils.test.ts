import 'isomorphic-fetch';
import * as sitemapUtils from '../../../src/middleware/sitemaps/sitemapUtils';
describe('createSitemapIndex()', () => {
  it('returns a sitemap index with no sitemaps', async () => {
    const sitemaps: sitemapUtils.SitemapSchemaSitemapElement[] = [];
    const res = sitemapUtils.createSitemapIndex(sitemaps);

    const xml = await res.text();

    expect(res.headers.get('Content-Type')).toBe('application/xml');
    expect(xml).toMatchSnapshot();
  });

  it('returns a sitemap index with specified sitemaps', async () => {
    const sitemaps: sitemapUtils.SitemapSchemaSitemapElement[] = [
      {
        loc: 'http://localhost:3000/post-sitemap.xml',
        lastmod: '2022-01-01',
      },
      {
        loc: 'http://localhost:3000/page-sitemap.xml',
      },
    ];

    const res = sitemapUtils.createSitemapIndex(sitemaps);

    const xml = await res.text();

    expect(res.headers.get('Content-Type')).toBe('application/xml');
    expect(xml).toMatchSnapshot();
  });
});

describe('createSitemap()', () => {
  it('returns a sitemap with no links', async () => {
    const urls: sitemapUtils.SitemapSchemaUrlElement[] = [];
    const res = sitemapUtils.createSitemap(urls);

    const xml = await res.text();

    expect(res.headers.get('Content-Type')).toBe('application/xml');
    expect(xml).toMatchSnapshot();
  });

  it('returns a sitemap with specified urls', async () => {
    const urls: sitemapUtils.SitemapSchemaUrlElement[] = [
      {
        loc: 'http://localhost:3000/',
        lastmod: '2022-01-01',
        changefreq: 'daily',
        priority: 0.9,
      },
      {
        loc: 'http://localhost:3000/about',
        lastmod: '2022-01-01',
      },
      {
        loc: 'http://localhost:3000/contact',
        priority: 0.4,
      },
      {
        loc: 'http://localhost:3000/team',
        changefreq: 'monthly',
      },
      {
        loc: 'http://localhost:3000/blog',
      },
    ];
    const res = sitemapUtils.createSitemap(urls);

    const xml = await res.text();

    expect(res.headers.get('Content-Type')).toBe('application/xml');
    expect(xml).toMatchSnapshot();
  });
});
