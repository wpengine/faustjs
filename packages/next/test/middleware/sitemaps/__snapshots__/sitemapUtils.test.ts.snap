// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSitemap() returns a sitemap with no links 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
    <urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\">
      
    </urlset>"
`;

exports[`createSitemap() returns a sitemap with specified urls 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
    <urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\">
      <url>
              <loc>http://localhost:3000/</loc>
              <lastmod>2022-01-01</lastmod>
              <changefreq>daily</changefreq>
              <priority>0.9</priority>
          </url><url>
              <loc>http://localhost:3000/about</loc>
              <lastmod>2022-01-01</lastmod>
              
              
          </url><url>
              <loc>http://localhost:3000/contact</loc>
              
              
              <priority>0.4</priority>
          </url><url>
              <loc>http://localhost:3000/team</loc>
              
              <changefreq>monthly</changefreq>
              
          </url><url>
              <loc>http://localhost:3000/blog</loc>
              
              
              
          </url>
    </urlset>"
`;

exports[`createSitemapIndex() returns a sitemap index with no sitemaps 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
      <sitemapindex xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\">
      
      </sitemapindex>"
`;

exports[`createSitemapIndex() returns a sitemap index with specified sitemaps 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
      <sitemapindex xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\">
      <sitemap>
            <loc>http://localhost:3000/post-sitemap.xml</loc>
            <lastmod>2022-01-01</lastmod>
          </sitemap><sitemap>
            <loc>http://localhost:3000/page-sitemap.xml</loc>
            
          </sitemap>
      </sitemapindex>"
`;
